sys/linux/bpf_trace.txt

```diff
@@ -0,0 +1,59 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <uapi/linux/bpf.h>
+
+# This file describes types of bpf programs that are attached using
+# BPF_RAW_TRACEPOINT_OPEN
+# These types are: BPF_PROG_TYPE_TRACING, BPF_PROG_TYPE_EXT, BPF_PROG_TYPE_LSM, BPF_PROG_TYPE_RAW_TRACEPOINT, BPF_PROG_TYPE_RAW_TRACEPOINT_WRITABLE
+
+resource bpf_lsm_btf_id[bpf_btf_id]
+resource fd_bpf_trace_prog[fd_bpf_prog]
+
+syz_btf_id_by_name$bpf_lsm(name ptr[in, string[bpf_lsm_func_names]]) bpf_lsm_btf_id (timeout[500])
+
+bpf$BPF_LSM_PROG_LOAD(cmd const[BPF_PROG_LOAD], arg ptr[in, bpf_lsm_prog], size len[arg]) fd_bpf_trace_prog (timeout[500])
+bpf$BPF_RAW_TRACEPOINT_PROG_LOAD(cmd const[BPF_PROG_LOAD], arg ptr[in, bpf_raw_tracepoint_prog], size len[arg]) fd_bpf_trace_prog
+bpf$BPF_TRACING_PROG_LOAD(cmd const[BPF_PROG_LOAD], arg ptr[in, bpf_tracing_prog], size len[arg]) fd_bpf_trace_prog
+bpf$BPF_RAW_TRACEPOINT_OPEN(cmd const[BPF_RAW_TRACEPOINT_OPEN], arg ptr[in, bpf_raw_tracepoint], size len[arg]) fd_perf_base (timeout[500])
+
+# NEED: After #2035 is merged (which we can't for now due to pahole issues),
+# syz_btf_id_by_name should work and we can remove the optional flag from
+# bpf_lsm_btf_id.
+type bpf_lsm_prog bpf_prog_t[const[BPF_PROG_TYPE_LSM, int32], const[BPF_LSM_MAC, int32], bpf_lsm_btf_id[opt], const[0, int32]]
+type bpf_raw_tracepoint_prog bpf_prog_t[flags[bpf_raw_tracepoint_prog_type, int32], const[0, int32], const[0, int32], const[0, int32]]
+type bpf_tracing_prog bpf_prog_t[flags[bpf_tracing_prog_type, int32], flags[bpf_tracing_attach_type, int32], int32[0:BTF_ID_MAX], fd_bpf_prog[opt]]
+
+bpf_raw_tracepoint_prog_type = BPF_PROG_TYPE_RAW_TRACEPOINT, BPF_PROG_TYPE_RAW_TRACEPOINT_WRITABLE
+bpf_tracing_prog_type = BPF_PROG_TYPE_TRACING, BPF_PROG_TYPE_EXT
+bpf_tracing_attach_type = BPF_TRACE_FENTRY, BPF_TRACE_FEXIT, BPF_TRACE_RAW_TP, BPF_MODIFY_RETURN, BPF_TRACE_ITER
+
+# An upper bound for reasonable btf_ids
+define BTF_ID_MAX	200000
```

Comment:
There is `BTF_MAX_TYPE` for this.

Reply:
I am aware of `BTF_MAX_TYPE`. However, `BTF_MAX_TYPE` is defined as ~= 1,000,000, but there are only ~ 160,000 IDs now and only 40000 of them are valid hook points. If we use `BTF_MAX_TYPE` I think the success rate will be quite low. 

With the kernel config Syzkaller currently used, there are ~163,000 btf_id in the latest stable kernel and ~163,600 in linux-next, so I think 200000 will be a suitable upper bound for at least a while.

Reply:
SGTM, but please add a comment explaining this.